Fix PodAntiAffinity issues in case of multiple affinityTerms

Huang-Wei opened the pull request:
- rename podMatchesAffinityTermProperties to podMatchesAllAffinityTermProperties
- add podMatchesAnyAffinityTermProperties which should be used in some PodAntiAffinity cases
- add unit tests

**What this PR does / why we need it**:

There are 2 phases to verify symmetry of PodAntiAffinity:
- verify labels of incoming pod not violates antiaffinity of existing pods
- verify antiaffinity of incoming pod not violates labels of existing pods

This PR is to fix issue in second step.

Currently logic of second step tries to invoke `podMatchesAffinityTermProperties` when checking both affinity and antiaffinity. 

It's not that accurate, as `podMatchesAffinityTermProperties` is a "match all" semantics, say for each term, ensure it matches target selector/labels, and return true if all term passes check. This is good for affinity, but not antiaffinity.

Regarding antiaffinity, it's a "match any" semantics - i.e. if found any term match target selector/labels, return true.

So without this fix, unit tests (which is added in this PR) failed on following cases:
- a) incoming pod's anti-affinity has more than one term - existing pod violates all terms
- b) incoming pod's anti-affinity has more than one term - existing pod violates partial term - case 1
- c) incoming pod's anti-affinity has more than one term - existing pod violates partial term - case 2

Case a) is guarded in before-mentioned step 1 of symmetric logic. So we don't see it in a running k8s cluster. Case b) and c) are real problems and specifically case c) is what issue #68101 reported.


**Which issue(s) this PR fixes**:
Fixes #68101

**Special notes for your reviewer**:
After reviewing the correctness of this PR, bsalamat and k82cn can decide if it needs to go into 1.12 release cycle. 

**Release note**:
```release-note
Fix PodAntiAffinity issues in case of multiple affinityTerms.
```

Huang-Wei commented:
/sig scheduling

Huang-Wei commented:
/cc @bsalamat @k82cn @ravisantoshgudimetla

Huang-Wei commented:
/retest

dixudx commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -356,8 +356,8 @@ func getAffinityTermProperties(pod *v1.Pod, terms []v1.PodAffinityTerm) (propert
 
-// podMatchesAffinityTermProperties return true IFF the given pod matches all the given properties.
-func podMatchesAffinityTermProperties(pod *v1.Pod, properties []*affinityTermProperties) bool {
+// podMatchesAllAffinityTermProperties return true if the given pod matches all the given properties.
```
IFF is not a typo.

Huang-Wei replied:
Got it. Updated.

resouer commented on pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -507,3 +507,235 @@ func TestPredicateMetadata_ShallowCopy(t *testing.T) {
+				Spec: v1.PodSpec{
+					Affinity: &v1.Affinity{
+						PodAntiAffinity: &v1.PodAntiAffinity{
+							RequiredDuringSchedulingIgnoredDuringExecution: newPodAffinityTerms([]string{"aaa", "ccc"}),
```
Existing pod lable: `"aaa": "", "bbb": ""`
New pod AntiAffinity is `"aaa", "ccc"`

It seems not `violates all terms`? Or I mis-read sth?

Huang-Wei replied:
Thanks for the catch. It's a copy/paste typo... Fixed.

resouer commented on pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -507,3 +507,235 @@ func TestPredicateMetadata_ShallowCopy(t *testing.T) {
+					},
+				},
+			},
+			wantAffinityPods:     make(map[string][]*v1.Pod),
```
You used nil above, should be consistent :D 

btw, I think nil is fine.

Huang-Wei replied:
Good catch, lol...

`getPodsMatchingAffinity` simply return nil if it doesn't have affinity/antiaffinity:

https://github.com/kubernetes/kubernetes/blob/058b26f38e49b64faad4a7a346b4616d0e9b57a0/pkg/scheduler/algorithm/predicates/metadata.go#L376-L382

But Mr. `reflect.DeepEqual` is unhappy to treat "nil" as equality of a "nil map" - so in some cases, I put nil, while some I put nil map there to make it happy...

BTW: `apiequality.Semantic.DeepEqual` is considerate to say yes to "nil equals to nil map". But maybe not worth introducing an extra package dependency.

resouer replied:
Ah I see, maybe just return empty object every time is enough.

Huang-Wei replied:
Hmm... Have to rebase code as PR #67788 refactored the data structure.

Not it's all consistent. PTAL.

resouer commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -369,6 +369,19 @@ func podMatchesAffinityTermProperties(pod *v1.Pod, properties []*affinityTermPro
 	return true
 }
 
+// podMatchesAnyAffinityTermProperties return true IFF the given pod matches any given property.
```
Should be `returns true ...`

It reads like this time, it's not a "if and only if" case? i.e.`It returns true if the given pod matches any given ...`?

Huang-Wei replied:
Agree. `It returns true if the given pod matches any given ...` should be clear enough for this "any" case.

(leave previous `IFF` in `podMatchesAllAffinityTermProperties` as is. although I think `if` is also clear there.)

resouer commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -356,8 +356,8 @@ func getAffinityTermProperties(pod *v1.Pod, terms []v1.PodAffinityTerm) (propert
 
-// podMatchesAffinityTermProperties return true IFF the given pod matches all the given properties.
-func podMatchesAffinityTermProperties(pod *v1.Pod, properties []*affinityTermProperties) bool {
+// podMatchesAllAffinityTermProperties return true IFF the given pod matches all the given properties.
```
Should be `returns true ...`

Huang-Wei replied:
Fixed.

resouer commented on pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -507,3 +507,235 @@ func TestPredicateMetadata_ShallowCopy(t *testing.T) {
+
+// TestGetPodsMatchingAffinity_AntiAffinity tests againt method getPodsMatchingAffinity
+// on Anti Affinity cases
+func TestGetPodsMatchingAffinity_AntiAffinity(t *testing.T) {
```
Maybe I am verbose, but I am thinking if you can add a test case to verify this bug is fixed in `predicates.go` as well?

Because we still have many logics there after we got the `antiAffinityPods` from metadata, so it would be great if we can ensure this change does work and will not break anything in `predicates.go`.

Sounds reasonable?

Huang-Wei replied:
@resouer Thanks for the detailed review!

Sure, it's totally reasonable. Do you think is it good enough to add additional UTs for `predicates.go#satisfiesPodsAffinityAntiAffinity()`? (step 2 of symmetry check this PR tries to fix)

resouer replied:
Yeah, should be.

Or, I think you can just add one more test case to verify this in `TestInterPodAffinity`? which seems quicker?

Huang-Wei replied:
Added one testcase - which only passed with this PR. PTAL.

resouer reviewed (commented):
The idea is valid to me, while some nits need to be fixed first :-) 

Also, I would suggest a test case for `predicates.go` to verify this fix solves the reported issue.

resouer commented on pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -507,3 +507,235 @@ func TestPredicateMetadata_ShallowCopy(t *testing.T) {
+			wantAntiAffinityPods: nil,
+		},
+		{
+			name: "non-affinity incoming pod causes a no-op", // the semmetric verification is covered elsewhere
```
It seems this comment is unnecessary? As elsewhere seems to be test cases below?

Huang-Wei replied:
Removed the comment to avoid confusion.

k82cn commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -458,7 +471,7 @@ func targetPodMatchesAffinityOfPod(pod, targetPod *v1.Pod) bool {
 		return false
 	}
-	return podMatchesAffinityTermProperties(targetPod, affinityProperties)
+	return podMatchesAllAffinityTermProperties(targetPod, affinityProperties)
```
Did not get time to go through all doc and comments :( 

But according to the comments of `PodAntiAffinity`, it said 

> 2309     // If the anti-affinity requirements specified by this field are not met at
2310     // scheduling time, the pod will not be scheduled onto the node.
2311     // If the anti-affinity requirements specified by this field cease to be met
2312     // at some point during pod execution (e.g. due to a pod label update), the
2313     // system may or may not try to eventually evict the pod from its node.
2314     // When there are multiple elements, the lists of nodes corresponding to each
2315     // podAffinityTerm are intersected, i.e. all terms must be satisfied.
2316     // +optional

And how do we handle Affinity that Node did not have such label? I'm thinking that a `bool` (Yes or No) maybe not enough :)

Huang-Wei replied:
"Node" is a typo I guess... I suppose you meant "pod".

If there is no pod has labels matching the terms, function `podMatches{All|Any}AffinityTermProperties (targetPod, affinityProperties)` both returns false.

This is enough I think, because:
> There are 2 phases to verify symmetry of PodAntiAffinity:
> - verify labels of incoming pod not violates {anti}affinity of existing pods
> - verify {anti}affinity of incoming pod not violates labels of existing pods 

In other words, we need to evaluate combination of each phase: (can do early break for sure)
- if it's Affinity case
    - only and only if `true` + `true` means fit
    - other combinations means no fit
- if it's AntiAffinity case
    - only and only if `false` + `false` means fit
    - other combinations means no fit

k82cn commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -428,11 +441,11 @@ func getPodsMatchingAffinity(pod *v1.Pod, nodeInfoMap map[string]*schedulercache
 			}
 			// Check anti-affinity properties.
-			if podMatchesAffinityTermProperties(existingPod, antiAffinityProperties) {
+			if podMatchesAnyAffinityTermProperties(existingPod, antiAffinityProperties) {
```
did we check `InterPodAffinityMatches` ?

Huang-Wei committed:
fix PodAntiAffinity issues

- update logic of verifying incoming pod's anti-affinity
- rename podMatchesAffinityTermProperties to podMatchesAllAffinityTermProperties
- add podMatchesAnyAffinityTermProperties which is used in some PodAntiAffinity cases
- rename some functions to make it more readable
- add unit tests to verify correctness of PodAffinity and PodAntiAffinity
  - verifying "Existing pod anti-affinity"
  - verifying "incoming pod's anti-affinity"
  - verifying "incoming pod's affinity"

Huang-Wei commented:
@resouer @k82cn Code is rebased and added testcase in `TestInterPodAffinity()`.

Please take another look. Thanks!

Huang-Wei commented:
/assign @bsalamat 
for 1.12 milestone consideration.

bsalamat commented:
> Regarding antiaffinity, it's a "match any" semantics - i.e. if found any term match target selector/labels, return true.

@Huang-Wei Where did you find this? Our API clearly say for both affinity and anti-affinity that "all terms must be satisfied":

https://github.com/kubernetes/kubernetes/blob/92ad24cc4dbd77d200e34f84a029d9bf282e8199/pkg/apis/core/types.go#L2319

Huang-Wei commented:
@bsalamat oh, sorry to confuse you. I'm not saying the API statement is incorrect. I just meant in the implementation, we can't use `podMatchesAffinityTermProperties` (which is ALL semantics) in both Affinity and AntiAffinity check:

> It's not that accurate, as podMatchesAffinityTermProperties is a "match all" semantics, say for each term, ensure it matches target selector/labels, and return true if all term passes check. This is good for affinity, but not antiaffinity.

> Regarding antiaffinity, it's a "match any" semantics - i.e. if found any term match target selector/labels, return true.

In other words, current behavior (without this fix) is: for antiaffinity, if and only if targetPod matches **all** terms of incoming pod, it **won't** be put onto target node.

While correct behavior (with this fix) is: for antiaffinity, if targetPod matches *any* term of incoming pod, it **won't** be put onto target node.

BTW: some detailed test cases are introduced in this PR.

bsalamat commented:
> While correct behavior (with this fix) is: for antiaffinity, if targetPod/existingPod matches any term of incoming pod, incoming pod won't be put onto target node.

No, that's not true. That's exactly what I am trying to ask. Where did you find the information that if "any term" matches, anti-affinity applies?

Huang-Wei commented:
> No, that's not true. That's exactly what I am trying to ask. Where did you find the information that if "any term" matches, anti-affinity applies?

It's kind of my understanding on "symmetry". A minimized example is like this (included in this PR in predicates_test.go):

```
- we have an existing pod on a node:
    - labels: {"b": ""}
    - PodAntiAffinity: term {"b" exists}
- then we got an incoming pod:
    - labels: {"a": ""}
    - PodAntiAffinity: term {"a" exists} and term {"b" exists}
```

With current logic, the incoming pod will be **approved** to land on the node b/c:
- it passed existingPod check: incoming labels: {"a": ""} not break existing PodAntiAffinity: term {"b" exists}
- it also passed incomingPod check: incoming PodAntiAffinity: term {"a" exists} and term {"b" exists} not break existing labels: {"b": ""}

But in my opinion, the second check is inaccurate - which is what this PR tries to fix.

And if we change the order of above example:

```
- we have an existing pod:
    - labels: {"a": ""}
    - PodAntiAffinity: term {"a" exists} and term {"b" exists}
- then we got an incoming pod:
    - labels: {"b": ""}
    - PodAntiAffinity: term {"b" exists}
```

In this case, incoming pod won't be **approved** b/c:

- it failed existingPod check: incoming labels: {"b": ""} violates existing PodAntiAffinity: term {"a" exists} and term {"b" exists}
- second check is bypassed

It's not symmetric.

Huang-Wei commented:
> In this case, the incoming pod must be approved as well. It doesn't match all the terms of the existing pod. My understanding is that it is approved with the existing logic, but if it is not, we should address it.

It kind of refreshes my understanding... But it's fine - we have the standard telling which is the right behavior. Thanks for pointing out @bsalamat .

To avoid confusion, I will go with another PR addressing it.

misterikkit commented on pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -525,3 +525,224 @@ func TestPredicateMetadata_ShallowCopy(t *testing.T) {
+			wantAntiAffinityPodsMaps: newTopologyPairsMaps(),
+		},
+		{
+			name: "non-affinity incoming pod causes a no-op",
```
What is "non-affinity"? I think the tests might be easier to read if you modify your `newPod` helper to create pods with or without affinity rules (or split it into multiple helpers) such that the test would read like,
```golang
existingPods: []*v1.Pod{newPod("p1"), newPodWithAffinity("p2", "aaa")},
```

Huang-Wei replied:
Description was updated to "incoming pod without affinity/anti-affinity causes a no-op".

misterikkit commented on pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -525,3 +525,224 @@ func TestPredicateMetadata_ShallowCopy(t *testing.T) {
+			wantAntiAffinityPodsMaps: newTopologyPairsMaps(),
+		},
+		{
+			name: "no pod has label violates incoming pod's anti-affinity",
```
grammar: s/no pod has label violates/no pod has label that violates/

misterikkit commented:
"PodAntiAffinity is not really symmetric in some cases" sounds more like an issue title than a PR title. Could you change the PR title and git commit message to describe the change, rather than the bug?

misterikkit commented:
> It ensures that all terms are satisfied, but this PR is trying to change it and make it "any term is satisfied".

I think there is some confusion over the meaning of "satisfy" and "matches". When I read @bsalamat's comments, it sounds like "**satisfied**" means, "(A) _all_ affinity terms must **match** and (B) _all_ anti-affinity terms must **not match**." The logic of this PR is trying to reject pods "if _any_ anti-affinity terms **match**" which is the contraposition of (B) above. (i.e. they are logically equivalent)

Following on from a couple of my code review comments, the naming and logic structure (and some innacurate code comments) seem to be causing confusion over whether this feature has been implemented correctly.

The fact that this PR adds and fixes an issue (https://github.com/kubernetes/kubernetes/issues/68101) should be examined more closely. Also, I would welcome PRs that add clarity to naming, comments, and structure.

Huang-Wei commented:
> I think there is some confusion over the meaning of "satisfy" and "matches". When I read @bsalamat's comments, it sounds like "satisfied" means, "(A) all affinity terms must match and (B) all anti-affinity terms must not match." The logic of this PR is trying to reject pods "if any anti-affinity terms match" which is the contraposition of (B) above. (i.e. they are logically equivalent)

@misterikkit thanks for the review! You're right this PR tries to reject pods "if any anti-affinity terms match". But I don't think Bobby meant this way.

For example, for following example:

```
- we have an existing pod on a node:
    - labels: {"b": ""}
    - PodAntiAffinity: term {"b" exists}
- then we got an incoming pod:
    - labels: {"a": ""}
    - PodAntiAffinity: term {"a" exists} and term {"b" exists}
```

I think incoming pod should be rejected as {term b} in incoming pod matches "labels b" in the existing pod.
But Bobby's idea is that {term a, term b} does not match all labels in the existing pod, so it should be an approval.

What's your opinion on this :)

Huang-Wei commented:
@bsalamat @misterikkit I found another analogy for our debate. The API doc says all terms should be ANDed.

- it's clear on semantic of Affinity: `AND (X, Y)` equals to `X AND Y` - we're all good on this agreement
- but it's ambiguous on AntiAffinity, does it mean
    1. `~ AND (X, Y)` which equals to `~X OR ~Y` , or
    1. `AND (~X, ~Y)` which equals to `~X AND ~Y`

PS: X, Y are AffinityTerm (we don't have AntiAffinityTerm). This PR assume case `i`.

Huang-Wei commented:
@bsalamat I will resolve all your comments once we have an agreement on what is the desired behavior. Thanks for your time again.

bsalamat commented:
@Huang-Wei The issue here is that you think that Anti-Affinity is Not of Affinity. That's not the case. Anti-affinity terms matching is the same as that of affinity. Just the final result is different. In other words two pods match one another's affinity (or anti-affinity) rules in the exact same way. If the rule is affinity, it means that the two pods should be scheduled in the same topology domain and if the rule is anti-affinity, it means that the two pods should not be scheduled in the same topology domain. Basically, the "NOT" applies to the placement strategy, not to the way terms are matched.

bsalamat commented:
@Huang-Wei That's right.

Huang-Wei commented:
@bsalamat Got it, that makes sense. Thanks for the patience...

Re:
> In this case, the incoming pod must be approved as well. It doesn't match all the terms of the existing pod. My understanding is that it is approved with the existing logic, but if it is not, we should address it.

Yes, it's a bug in 1.12 and 1.11. I will open another PR addressing it.

resouer commented:
@bsalamat Thanks a lot for detailed explanation! Many users should misunderstand the case indeed.

@Huang-Wei Upon this, I think it deserves a mention in the doc repo as well.

bsalamat commented:
I agree with @resouer that we should improve our documentation with more details and examples.

> The currently logic which use label selector to match pod is suspicious BTW.

What do you mean by suspicious? You think that it may have a bug or you think the logic is broken?

Huang-Wei commented:
@bsalamat I think Harry meant it's a bug. And I agree with that. That line instantly built a topologyPair and put into topologyPairMap when one term matches. But actually we should wait for all terms' match.

bsalamat commented:
> I think Harry meant it's a bug. And I agree with that. That line instantly built a topologyPair and put into topologyPairMap when one term matches. But actually we should wait for all terms' match.

Yes, that's right. It is a bug and we should fix it.

Huang-Wei commented:
@bsalamat yeap, I'm working on it.

Huang-Wei commented:
This PR gives fix incorrectly but has positive discussion and clarification.

Correct PR is put in #68301. Close this PR.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -439,7 +452,7 @@ func targetPodMatchesAffinityOfPod(pod, targetPod *v1.Pod) bool {
 		glog.Errorf("error in getting affinity properties of Pod %v", pod.Name)
 		return false
 	}
```
done.

Huang-Wei commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

Huang-Wei commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1242,7 +1242,11 @@ func GetPodAntiAffinityTerms(podAntiAffinity *v1.PodAntiAffinity) (terms []v1.Po
+// getTPMapMatchingExistingAntiAffinity calculates the following for each existing pod on each node:
+// (1) Whether it has PodAntiAffinity
+// (2) Whether any AffinityTerm matches the incoming pod
+// TODO(Huang-Wei): move this to metadta.go?
```
considering that function is only invoked in `metadata.go` as an initilization helper, and similar function `getTPMapMatchingIncomingAffinityAntiAffinity()` is also defined/called in `metadata.go`. I think we should move it there as well.

Huang-Wei replied:
I'll move this function before squashing commits. As moving a function usually results in a tricky `diff` view.

Huang-Wei commented:
/cc @bsalamat @resouer @misterikkit for another round of review.

After discussing with @bsalamat on the right behavior of AntiAffinity - esp. on the case of multi-terms. We've finally went to a [consensus](https://github.com/kubernetes/kubernetes/pull/68301#discussion_r216834596):

> Existing pod anti-affinity: If the incoming pod matches the label selector and namespace of any anti-affinity term of an existing pod, the topologyKey and the corresponding label value of the node should be added to the topology pair map. The reason is that this term would be violated if the incoming pod is scheduled on in that topology. So, here we don't need to check all terms. Any term that is matched, should be added. (Our implementation before this PR was valid)

> Incoming pod anti-affinity: If any existing pod on a node matches any terms of the incoming Pod anti-affinity, the topology key and its corresponding value on the node should be added to the map of topology pairs. This is similar to the above rule.

> Incoming pod affinity: This is where things are different. For an affinity rule to be considered satisfied, all the terms of the affinity rule must match. Based on a previous decision, a single pod (as opposed to a group of pods) must match all the terms for the affinity rule to be considered satisfied. I don't have a great idea on how we can use our topology pair map for checking pod affinity efficiently. I think that the implementation that we had in #62211 would work better for matching affinity, but I might be wrong.

Latest commit gives fixes on buggy behavior of `Incoming pod anti-affinity`, as well as a number of unit tests, new comments, and function renamings we discussed in #68301.

Huang-Wei commented:
/test pull-kubernetes-kubemark-e2e-gce-big

Huang-Wei commented:
@bsalamat I didn't see any expected behavior before/after this PR regarding "Existing pod anti-affinity" and "Incoming pod affinity". Please share if you have any "suspicious" case in mind.

k82cn commented:
/kind bug

bsalamat commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1322,26 +1313,24 @@ func getMatchingTopologyPairsOfExistingPod(newPod *v1.Pod, existingPod *v1.Pod,
 }
-func (c *PodAffinityChecker) getMatchingAntiAffinityTopologyPairs(pod *v1.Pod, allPods []*v1.Pod) (*topologyPairsMaps, error) {
+
+func (c *PodAffinityChecker) calculateMatchingAntiAffinityTopologyPairs(pod *v1.Pod, allPods []*v1.Pod) (*topologyPairsMaps, error) {
```
To be consistent, please rename this to `getMatchingAntiAffinityTopologyPairsOfAllPods`.

Huang-Wei replied:
done.

bsalamat commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1400,6 +1389,9 @@ func (c *PodAffinityChecker) nodeMatchesTopologyTerms(pod *v1.Pod, topologyPairs
 				return false
 			}
 		} else {
+			// TODO(Huang-Wei): if incoming pod has an anti-affinity with invalid topology
```
We should have two functions, instead of this one: `nodeMatchesAllTopologyTerms` and `nodeMatchesAnyTopologyTerms`. We should then use `nodeMatchesAllTopologyTerms` for checking affinity and `nodeMatchesAnyTopologyTerms` for checking anti-affinity.

Huang-Wei replied:
Good point. Done.

bsalamat commented:
@Huang-Wei Given the size of this PR, please commit new changes in separate commits (do not amend existing ones). We will squash them when the PR is ready.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -408,7 +425,7 @@ func getPodsMatchingAffinity(pod *v1.Pod, nodeInfoMap map[string]*schedulercache
 			// Check anti-affinity properties.
-			if podMatchesAffinityTermProperties(existingPod, antiAffinityProperties) {
+			if podMatchesAnyAffinityTermProperties(existingPod, antiAffinityProperties) {
 				for _, term := range antiAffinityTerms {
```
Right. Done.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -338,11 +342,24 @@ func podMatchesAffinityTermProperties(pod *v1.Pod, properties []*affinityTermPro
+}
+
+// getTPMapMatchingIncomingAffinityAntiAffinity finds existing Pods that match affinity terms of the given "pod".
+// It ignores topology. It returns a topologyPairsMaps that are checked later by the affinity
```
done

Huang-Wei commented on pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -3171,12 +3171,15 @@ func TestInterPodAffinityWithMultipleNodes(t *testing.T) {
+				"nodeB": false,
 			},
-			name: "This test ensures that anti-affinity matches a pod when all terms of the anti-affinity rule matches a pod.",
+			name: "This test ensures that anti-affinity matches a pod when any term of the anti-affinity rule matches a pod.",
```
In this test, nodeB can't be a fit as it's in "region r1".

If it's put to nodeB, it violates anti-affinity terms of incoming pod - which doesn't want to put into "region r1" which has pod with label "foo in abc". (where existing pod does, although it's in another node)

@bsalamat please double check and verify it's fine.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -3724,10 +3781,11 @@ func TestInterPodAffinityWithMultipleNodes(t *testing.T) {
 			fits: map[string]bool{
 				"nodeA": false,
-				"nodeB": true,
+				"nodeB": false,
```
the change is the same as https://github.com/kubernetes/kubernetes/pull/68173/commits/d20e5e3e699cf927ba29552f7b735159cbe868d5#r217482459

Huang-Wei replied:
and I can delete this case as it's duplicate as https://github.com/kubernetes/kubernetes/pull/68173/commits/d20e5e3e699cf927ba29552f7b735159cbe868d5#r217482459

Huang-Wei commented on pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -3624,6 +3627,60 @@ func TestInterPodAffinityWithMultipleNodes(t *testing.T) {
+				"nodeA": false,
+				"nodeB": true,
+			},
+			name: "Test incoming pod's anti-affinity: only when labelSelector and topologyKey both matches, it's counted as a single term match - case when one term has invalid topologyKey",
```
We have a consistent behavior now: no matter it's affinity or anit-affinity, no matter it's multi-terms or single term, no matter it's partial or full match on terms, no matter it's partial or full match on topologyKey, and also no matter there is invalid topologyKey or not.

Huang-Wei commented:
@bsalamat please take a look on new commit.

bsalamat commented on pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -525,3 +525,224 @@ func TestPredicateMetadata_ShallowCopy(t *testing.T) {
+
+// TestGetTPMapMatchingIncomingAffinityAntiAffinity_AntiAffinity tests against method getTPMapMatchingIncomingAffinityAntiAffinity
+// on Anti Affinity cases
+func TestGetTPMapMatchingIncomingAffinityAntiAffinity_AntiAffinity(t *testing.T) {
```
The "_AntiAffinity" suffix seems to be unintentional. Please remove.

Huang-Wei replied:
Done.

bsalamat commented on pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -525,3 +525,224 @@ func TestPredicateMetadata_ShallowCopy(t *testing.T) {
+		existingPods             []*v1.Pod
+		nodes                    []*v1.Node
+		pod                      *v1.Pod
+		wantAffinityPodsMaps     *topologyPairsMaps
```
Looks like none of the test cases below exercise Affinity maps. You may want to add a couple of affinity cases.

Huang-Wei replied:
Done, as well as revising the whole testing function.

bsalamat commented on pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -2546,6 +2546,349 @@ func TestInterPodAffinity(t *testing.T) {
+			node:                 &node1,
+			fits:                 false,
+			expectFailureReasons: []algorithm.PredicateFailureReason{ErrPodAffinityNotMatch, ErrExistingPodsAntiAffinityRulesNotMatch},
+			name:                 "PodAntiAffinity symmetry check a2: incoming pod and existing pod partially match each other on AffinityTerms",
```
Can we have a test case similar to this one in `TestInterPodAffinityWithMultipleNodes` which multiple nodes. Test scenario:
- We have two nodes, each running a pod.
- Each one of the existing pods has anti-affinity with multiple terms (two is enough).
- An incoming pod matches one of the anti-affinity terms of existing on each node and therefore, should not be schedulable.
- we can optionally add a third node and verify that the pod is schedulable there.

Huang-Wei replied:
Done. It's added in latest commit in `TestInterPodAffinityWithMultipleNodes()`.

bsalamat reviewed (commented):
Thanks, @Huang-Wei! The logic looks correct to me. I have a couple comments about the tests.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1273,81 +1277,68 @@ func getMatchingTopologyPairs(pod *v1.Pod, nodeInfoMap map[string]*schedulercach
 }
-func (c *PodAffinityChecker) getMatchingAntiAffinityTopologyPairs(pod *v1.Pod, allPods []*v1.Pod) (*topologyPairsMaps, error) {
+
+func (c *PodAffinityChecker) getMatchingAntiAffinityTopologyPairsOfAllPods(pod *v1.Pod, allPods []*v1.Pod) (*topologyPairsMaps, error) {
```
Done.

Huang-Wei commented:
@bsalamat Thanks for the review! PTAL.

Huang-Wei commented:
/retest

bsalamat reviewed (commented):
Thanks, @Huang-Wei! LGTM. Please squash commits. We are ready to merge.

bsalamat reviewed (approved):
/lgtm

Huang-Wei commented:
Thanks @bsalamat ! Commits have been squashed.

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

rphillips commented:
/xref https://github.com/kubernetes-incubator/bootkube/issues/1001

The changes in this PR are causing `affinity` sections to appear in checkpoint files causing the Kubelet to not restore pods. The `affinity` block within the checkpoint manifest errors with the following on kubelet restart:

```
predicate.go:133] Predicate failed on Pod: pod-checkpointer-l9cgg-172.17.4.101_kube-system(231cd4bc9c8f63b3131c1ec25716fe91), for reason: Predicate MatchNodeSelector failed
```

Checkpoint manifest json includes:
```
"affinity": {
      "nodeAffinity": {
        "requiredDuringSchedulingIgnoredDuringExecution": {
          "nodeSelectorTerms": [
            {
              "matchExpressions": null
            }
          ]
        }
      }
    },
```

Removing the affinity section within the checkpoint manifest fixes the restore issue. I'll be investigating the issue further, but am commenting to see if the upstream authors have any thoughts. This issue is a breaking change on upgrades and new installs for 1.12.x.

rphillips commented:
I suspect setting the affinity to nil when matchExpressions == null/nil would preserve the prior behavior.

Huang-Wei commented:
@rphillips thanks for bringing this up.

If I understand correctly, you're suspicious that after applying this PR, scheduler rejects a pod which has `nil` matchExpression in nodeAffinity?

Huang-Wei commented:
Considering this PR doesn't go to v1.11.3, I did a test on v1.11.3:

```yaml
kind: Pod
apiVersion: v1
metadata:
  name: pause
spec:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions: null
  containers:
  - name: pause
    image: k8s.gcr.io/pause:3.1
```

The above pod is also rejected. In other words, this PR seems not to break behavior of null matchExpressions.

```console
$ k get po
NAME      READY     STATUS    RESTARTS   AGE
pause     0/1       Pending   0          2m
$ k describe po pause | tail -3
  Type     Reason            Age               From               Message
  ----     ------            ----              ----               -------
  Warning  FailedScheduling  1m (x25 over 3m)  default-scheduler  0/3 nodes are available: 3 node(s) didn't match node selector.
```

rphillips commented:
@Huang-Wei Thank you for the prompt reply. The concern is with a static pod the kubelet will restore on restart. Bootkube, Tectonic, and other distributions may use a checkpointer to save the state in a manifest.

```
core@c1 /etc/kubernetes/manifests $ jq < kube-system-pod-checkpointer-q8zww.json 
{
  "kind": "Pod",
  "apiVersion": "v1",
  "metadata": {
    "name": "pod-checkpointer-q8zww",
    "namespace": "kube-system",
    "creationTimestamp": null,
    "labels": {
      "controller-revision-hash": "6775765d66",
      "k8s-app": "pod-checkpointer",
      "pod-template-generation": "1",
      "tier": "control-plane"
    },
    "annotations": {
      "checkpointer.alpha.coreos.com/checkpoint-of": "pod-checkpointer-q8zww"
    },
    "ownerReferences": [
      {
        "apiVersion": "v1",
        "kind": "Pod",
        "name": "pod-checkpointer-q8zww",
        "uid": "55c42d8e-cd79-11e8-9476-080027385a9b",
        "controller": true
      }
    ]
  },
  "spec": {
    "volumes": [
      {
        "name": "kubeconfig",
        "hostPath": {
          "path": "/etc/kubernetes/checkpoint-configmaps/kube-system/pod-checkpointer-q8zww/kubeconfig-in-cluster"
        }
      },
      {
        "name": "etc-kubernetes",
        "hostPath": {
          "path": "/etc/kubernetes",
          "type": ""
        }
      },
      {
        "name": "var-run",
        "hostPath": {
          "path": "/var/run",
          "type": ""
        }
      },
      {
        "name": "pod-checkpointer-token-ljvdc",
        "hostPath": {
          "path": "/etc/kubernetes/checkpoint-secrets/kube-system/pod-checkpointer-q8zww/pod-checkpointer-token-ljvdc"
        }
      }
    ],
    "containers": [
      {
        "name": "pod-checkpointer",
        "image": "quay.io/coreos/pod-checkpointer:9dc83e1ab3bc36ca25c9f7c18ddef1b91d4a0558",
        "command": [
          "/checkpoint",
          "--lock-file=/var/run/lock/pod-checkpointer.lock",
          "--kubeconfig=/etc/checkpointer/kubeconfig",
          "--checkpoint-grace-period=5m"
        ],
        "env": [
          {
            "name": "NODE_NAME",
            "valueFrom": {
              "fieldRef": {
                "apiVersion": "v1",
                "fieldPath": "spec.nodeName"
              }
            }
          },
          {
            "name": "POD_NAME",
            "valueFrom": {
              "fieldRef": {
                "apiVersion": "v1",
                "fieldPath": "metadata.name"
              }
            }
          },
          {
            "name": "POD_NAMESPACE",
            "valueFrom": {
              "fieldRef": {
                "apiVersion": "v1",
                "fieldPath": "metadata.namespace"
              }
            }
          }
        ],
        "resources": {},
        "volumeMounts": [
          {
            "name": "kubeconfig",
            "mountPath": "/etc/checkpointer"
          },
          {
            "name": "etc-kubernetes",
            "mountPath": "/etc/kubernetes"
          },
          {
            "name": "var-run",
            "mountPath": "/var/run"
          },
          {
            "name": "pod-checkpointer-token-ljvdc",
            "readOnly": true,
            "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount"
          }
        ],
        "terminationMessagePath": "/dev/termination-log",
        "terminationMessagePolicy": "File",
        "imagePullPolicy": "Always"
      }
    ],
    "restartPolicy": "Always",
    "terminationGracePeriodSeconds": 30,
    "dnsPolicy": "ClusterFirst",
    "nodeSelector": {
      "node-role.kubernetes.io/master": ""
    },
    "nodeName": "172.17.4.101",
    "hostNetwork": true,
    "securityContext": {},
    "affinity": {
      "nodeAffinity": {
        "requiredDuringSchedulingIgnoredDuringExecution": {
          "nodeSelectorTerms": [
            {
              "matchExpressions": null
            }
          ]
        }
      }
    },
    "schedulerName": "default-scheduler",
    "tolerations": [
      {
        "key": "node-role.kubernetes.io/master",
        "operator": "Exists",
        "effect": "NoSchedule"
      },
      {
        "key": "node.kubernetes.io/not-ready",
        "operator": "Exists",
        "effect": "NoExecute"
      },
      {
        "key": "node.kubernetes.io/unreachable",
        "operator": "Exists",
        "effect": "NoExecute"
      },
      {
        "key": "node.kubernetes.io/disk-pressure",
        "operator": "Exists",
        "effect": "NoSchedule"
      },
      {
        "key": "node.kubernetes.io/memory-pressure",
        "operator": "Exists",
        "effect": "NoSchedule"
      },
      {
        "key": "node.kubernetes.io/unschedulable",
        "operator": "Exists",
        "effect": "NoSchedule"
      },
      {
        "key": "node.kubernetes.io/network-unavailable",
        "operator": "Exists",
        "effect": "NoSchedule"
      }
    ],
    "priority": 0
  },
  "status": {}
}
```

rphillips commented:
Note: We do not create the checkpointer pod with an affinity, so I'm not sure where it is being added in.

Huang-Wei commented:
> We do not create the checkpointer pod with an affinity, so I'm not sure where it is being added in.

Agree, this is the key point.
